Allow editing time entries

shantanuraj opened the pull request:
## What does this PR do?

Adds editor-based editing of time entries on top of the flag-based path that shipped in #97.

```shell
$ toggl edit                                  # opens the running entry in $EDITOR
$ toggl edit <id>                             # opens that entry in $EDITOR
$ toggl edit --description "new desc"         # flag-based, unchanged behavior
$ toggl edit <id> --project "Toggl"           # flag-based on a specific entry
```

If no edit flags are passed, the entry is serialized to a plain-text file and opened in `$VISUAL`, then `$EDITOR`, falling back to `vim`. The buffer is read back when the editor exits, parsed, project/task names are resolved against the workspace, and the entry is updated. With any of `--description` / `--project` / `--tags`, the flag-based path from #97 runs unchanged.

Selection without an explicit id: the running entry if there is one, otherwise the most recent entry.

The core is a small `Parcel` trait:

```rust
pub trait Parcel {
    fn serialize(&self) -> Vec<u8>;
    fn deserialize(data: Vec<u8>, base: &Self) -> ResultWithDefaultError<Self>
    where
        Self: Sized;

    fn update_in_editor(&self) -> ResultWithDefaultError<Self> { /* tempfile + editor */ }
}
```

The buffer format is line-oriented `Key: Value`. Setting a value to empty clears that field. Bad input surfaces as a typed error with the offending line — no panics.

### Changes

- 🎬 `toggl edit` opens the entry in `$EDITOR` when no edit flags are given
- 🪡 Editor fallback chain: `VISUAL` → `EDITOR` → `vim`
- 💸 `Billable` is part of the buffer
- 🧷 Selection falls back to most recent entry when nothing is running
- 🧱 `Parcel` trait with proper error propagation (replaces the earlier `.expect`/`.unwrap` flow)

## Todos

- [x] Allow updating time-entry using editor
- [x] Allow editing running time-entry
- [x] Allow editing last time-entry (falls back to most recent when nothing is running)
- [x] Allow editing specific time-entry (`toggl edit <id>`)
- [x] Handle billable correctly
- [x] Accept time-entry edits from flags (shipped in #97)
- [x] Nicer editor fallback (`VISUAL`, then `EDITOR`, then `vim`)
- [x] Handle a line being removed correctly, ie. unset the field (deferred — empty value works today, removing the whole line keeps the prior value)
- [ ] Implement interactive picker for selecting time-entry to edit (deferred)
- [ ] Maybe add `editor` configuration to `toggl config` (deferred)

### Context

Builds on top of #97. The two designs coexist: flag-based for scripted edits, editor-based for the longer-form case (description rewrites, fixing the start time, etc.).

---

[^1]: Editor resolution: `\$VISUAL` then `\$EDITOR` then `vim`. `TOGGL_EDITOR` can be revisited if there's appetite to override `VISUAL/EDITOR` only for this CLI.

wd60622 commented:
I haven't done any programming in rust (yet) but I am building from source and will give it a try!

shantanuraj commented:
That's great! If you run into any issues feel to reach out.

wd60622 commented:
Would it make sense to also support flags here for non-interactive mode. For instance, `toggl edit --description "new description"` would change the description.

Only `toggl edit --interactive` would open the editor in my mind. What do you think?

shantanuraj commented:
Yeah at the moment that's the case, and I feel 90% of the time that is how it will end
up being used.

I feel it'd be nice if `edit` works for for any entry so you could use the cli to list entries
and you if see something out of place you can correct it from within the cli.

Though if you feel strongly in either direction let us know!
I'm quite open to changing how this works.

shantanuraj commented:
@wd60622 I've added a Todos list to the PR description, and pushed some changes so editing the running entry now works with the editor of your choice and `toggl edit`.
You can update description, tags, project, task and start time for now.

wd60622 commented:
Hi @shantanuraj , back looking at this again 👋 

Definitely think approaching 1. current time entry and 2. all past time entries separately is good idea. If current time entry is easiest, then that should be first. 

Thoughts on the difficulty to  make it non-interactive. For instance, `toggl edit --description <new-description>` or `toggl edit --start-time <new-start-time>`

shantanuraj committed:
🎬 Allow editing time entries via $EDITOR

- `toggl edit` with no edit flags opens the entry in `$EDITOR`
  - Falls back through `$VISUAL` → `$EDITOR` → `vim`
- With any of `--description` / `--project` / `--tags`, stays on the
  flag-based path from #97
- Selection without an explicit id picks the running entry, falling
  back to the most recent one if nothing is running

The core is a `Parcel` trait that writes the entry to a temp file as
`Key: Value` lines, spawns the editor, and reads it back.

```rust
pub trait Parcel {
    fn serialize(&self) -> Vec<u8>;
    fn deserialize(data: Vec<u8>, base: &Self) -> ResultWithDefaultError<Self>
    where
        Self: Sized;
}
```

Note: Bad input surfaces as a typed error with the offending line.
Previously similar paths used `.unwrap()` / `.expect()` and panicked.

shantanuraj committed:
✂️ Treat removed lines as field unsets in the editor buffer

- Buffer is the truth, any field whose line is removed (or whose
  value is left blank) is cleared on save
- Required scaffolding falls back to the original entry
  - `id`
  - `start`
  - `workspace_id`
  - `duration`
  - `created_with`

Previously a missing line left the prior value in place, so deleting
the `Tags` line silently kept the old tags.

shantanuraj committed:
⏱️ Recompute duration after editor save

- `start` and `stop` are user-editable in the buffer, `duration` is
  derived from them
- After deserialize, set `duration` based on the final state
  - `stop` set: `(stop - start).num_seconds()`
  - `stop` cleared: `-start.timestamp()` (running marker)
- Matches the formula used by `as_running_time_entry` and
  `as_stopped_time_entry`

shantanuraj committed:
🏷️ Always send tags as an array so empty actually clears

- Toggl's go API handles `null` as omitempty.
- The TimeEntry → NetworkTimeEntry conversion was producing `null`
  for an empty Vec, so removing all tags in the editor silently
  left them in place.
- Always send Some(vec), empty becomes `[]`.

shantanuraj committed:
🕰️ Add `--start-time` and `--stop-time` to `toggl edit`

- Both flags expect an RFC3339 timestamp, e.g. `2026-05-11T07:39:45Z`
- `--stop-time ""` clears the stop and marks the entry running again,
  matching the `--project ""` / `--tags ""` convention
- `duration` is recomputed from the final start/stop
  - stopped: `(stop - start).num_seconds()`
  - running: `-start.timestamp()`
